From 09870d2772b284d0061a5e4d1e1cdf6fb6764344 Mon Sep 17 00:00:00 2001
From: Dongsheng Yang <yangds.fnst@cn.fujitsu.com>
Date: Tue, 11 Nov 2014 07:18:22 -0500
Subject: [PATCH] btrfs: qgroup: return EINVAL if level of parent is not higher
 than child's.
Git-commit: 09870d2772b284d0061a5e4d1e1cdf6fb6764344
Patch-mainline: v4.1
References: bsc#972951

When we create a subvol inheriting a qgroup, we need to check the level
of them. Otherwise, there is a chance a qgroup can inherit another qgroup
at the same level.

Signed-off-by: Dongsheng Yang <yangds.fnst@cn.fujitsu.com>
Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Mark Fasheh <mfasheh@suse.de>
---
 fs/btrfs/qgroup.c | 5 +++++
 1 file changed, 5 insertions(+)

--- a/fs/btrfs/qgroup.c
+++ b/fs/btrfs/qgroup.c
@@ -1704,6 +1704,11 @@ int btrfs_qgroup_inherit(struct btrfs_tr
 				ret = -EINVAL;
 				goto out;
 			}
+
+			if ((srcgroup->qgroupid >> 48) <= (objectid >> 48)) {
+				ret = -EINVAL;
+				goto out;
+			}
 			++i_qgroups;
 		}
 	}
